MapboxVision/Services/VideoPlayer.swift

```diff
@@ -0,0 +1,148 @@
+//
+//  VideoPlayer.swift
+//  MapboxVision
+//
+//  Created by Alexander Pristavko on 5/17/19.
+//  Copyright © 2019 Mapbox. All rights reserved.
+//
+
+import Foundation
+import CoreMedia
+import AVFoundation
+
+protocol VideoPlayable: VideoSource {
+    var delegate: VideoPlayerDelegate? { get set }
+
+    func start()
+    func stop()
+}
+
+protocol VideoPlayerDelegate: AnyObject {
+    func playbackDidFinish()
+}
+
+private enum PlaybackConstants {
+    static let frameDuration: TimeInterval = 0.03
+}
+
+final class VideoPlayer: NSObject, VideoPlayable {
+    weak var delegate: VideoPlayerDelegate?
+
+    private var isPlaying = false
+    private let player: AVPlayer
+    private let videoOutput: AVPlayerItemVideoOutput
+    private var displayLink: CADisplayLink!
+
+    private let observers = ObservableVideoSource()
+    private var notificationToken: NSObjectProtocol?
+
+    private let queue = DispatchQueue(label: "com.mapbox.VideoPlayer")
+
+    init(path: String) throws {
+        guard FileManager.default.fileExists(atPath: path) else {
+            throw CocoaError(.fileNoSuchFile)
+        }
+
+        let url = URL(fileURLWithPath: path)
+        let asset = AVAsset(url: url)
+        let playerItem = AVPlayerItem(asset: asset)
+
+        player = AVPlayer(playerItem: playerItem)
+        player.actionAtItemEnd = .none
+
+        let attributes = [String(kCVPixelBufferPixelFormatTypeKey) : Int(kCVPixelFormatType_32BGRA)]
+        videoOutput = AVPlayerItemVideoOutput(pixelBufferAttributes: attributes)
+
+        super.init()
+
+        videoOutput.setDelegate(self, queue: queue)
+        playerItem.add(videoOutput)
+
+        self.displayLink = CADisplayLink(target: self, selector: #selector(updateOnDisplayLink))
+        displayLink.add(to: RunLoop.current, forMode: RunLoop.Mode.default)
+        displayLink.isPaused = true
+
+        notificationToken = NotificationCenter.default.addObserver(
+            forName: Notification.Name.AVPlayerItemDidPlayToEndTime,
+            object: playerItem,
+            queue: .main
+        ) { [weak self] _ in
+            self?.stop()
+            self?.delegate?.playbackDidFinish()
```

Comment:
Not sure if `playbackDidFinish` is properly synced with `currentItem?.seek(to: .zero, completionHandler: nil)` especially in case `seek()` method failed.

Reply:
Probably need to review the `AVPlayerItemDidPlayToEndTime` trigger and a chain of triggers after.

Reply:
Let's review and create separate issue.